MAINT: support python 3.10

tacaswell opened the pull request:
In https://github.com/python/cpython/pull/20290 CPython changed
`Py_TYPE` from a macro to an inline function.  This requires a code
change to us `Py_SET_TYPE` instead when using `Py_TYPE()` as a lvalue
in c code.

~This is a draft (and I expect it to fail all of the tests) because `Py_SET_TYPE` was only introduced in py39.  Can I please have some guidance on how to do version gating (or have someone who knows take over this PR ;) )~

tacaswell committed:
MNT: support python 3.10

In https://github.com/python/cpython/pull/20290 CPython changed
`Py_TYPE` from a macro to an inline function.  This requires a code
change to us `Py_SET_TYPE` instead when using `Py_TYPE()` as a lvalue
in c code.

In https://github.com/python/cpython/pull/20429 CPython changed
`Py_SIZE` from a macro to an inline function.  This requires a code
change to us `Py_SET_SIZE` instead of using `Py_SIZE` as a lvalue in c
code.

charris commented:
Ugh. Is it certain that this change will persist to release?

eric-wieser commented:
Something like this in `npy3k.h` (or however it's spelt) ]would do it:

```
#if HEX_VERSION < 0x031000
    #define Py_SET_TYPE(obj, typ) (Py_TYPE(obj) = typ)
#endif
```

tacaswell commented:
This is part of a series of intentional API breaks to try and condense the python c-api so I am pretty confident that this will persist.

seberg commented:
It seems like we should also replace all `ob_type` then... But, I guess no need to try and get ahead of Python here...

charris commented:
> This is part of a series of intentional API breaks

We might want to make a compatibility include file to use until Python3.9 gets dropped.  Probably need to use new names to avoid redefinitions.

tacaswell committed:
MNT: define macros supplied in py39 and above

To support the change to Py_SIZE and Py_TYPE the Py_SET_SIZE and
Py_SET_TYPE macros are provided in py39.  This provides the same
macros falling back to the old method for < py39.

eric-wieser commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -546,4 +546,9 @@ NpyCapsule_Check(PyObject *ptr)
 }
 #endif
 
+// account for changing PyType
+#if PY_VERSION_HEX < 0x03090000
```
```suggestion
/* Introduced in https://github.com/python/cpython/commit/d2ec81a8c99796b51fb8c49b77a7fe369863226f */
#if PY_VERSION_HEX < 0x030900a4
```

charris commented:
We are running tests against 3.9-dev, so that should validate that this works.

tacaswell committed:
MNT: be more precise about version support

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

charris commented on numpy/core/include/numpy/npy_3kcompat.h:
```diff
@@ -546,4 +546,10 @@ NpyCapsule_Check(PyObject *ptr)
 }
 #endif
 
+/* Introduced in https://github.com/python/cpython/commit/d2ec81a8c99796b51fb8c49b77a7fe369863226f */
```
This might look better up top with the similar compatibility hack.

tacaswell replied:
I should also add a reference for the other one.

charris commented:
The testing error is on the azure end and may be ignored.

charris commented:
Thanks Thomas.
